### PR TITLE
Add VBMSError and parent DependencyError classes

### PR DESCRIPTION
**Pull request by nemodjuric**
Filter for VBMS errors, similar (but different) what we have in Caseflow (see https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/exceptions/vbms_error.rb)

### CODE REVIEW

File: app/services/external_api/vbms_service.rb [hunk 1]
```diff
@@ -33,21 +33,11 @@ def self.v2_fetch_documents_for(source)
...
-      raise unless e.body.include?("File Number does not exist within the system.")
-
-      alternative_file_number = ExternalApi::BGSService.new.fetch_veteran_info(veteran_file_number)["file_number"]
-
-      raise if alternative_file_number == veteran_file_number
-
-      request = VBMS::Requests::FindDocumentVersionReference.new(alternative_file_number)
-      documents = send_and_log_request(alternative_file_number, request)
-    end
-
```
Comment (nemodjuric) [inline app/services/external_api/vbms_service.rb]: This became obsolete once we started using always the claim number (or vbms id). See https://github.com/department-of-veterans-affairs/caseflow-efolder/pull/1060

File: app/services/external_api/vbms_service.rb [hunk 2]
```diff
@@ -64,6 +54,8 @@ def self.v2_fetch_document_file(document)
     request = VBMS::Requests::GetDocumentContent.new(document.document_id)
     result = send_and_log_request(document.document_id, request)
     result&.content
+  rescue StandardError => e
+    raise ::VBMSError.from_dependency_error(e)
```
Comment (nemodjuric) [inline app/services/external_api/vbms_service.rb]: Similarly to bgs_service, not all methods are used in vbms_service.

File: app/exceptions/vbms_error.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+# Wraps known BGS errors so that we can better triage what gets reported in Sentry alerts.
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: comment refers to BGS

File: app/exceptions/vbms_error.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    /HTTPClient::KeepAliveDisconnected:/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/2847/
+    /HTTPClient::ConnectTimeoutError: exection expired/ => "TransientVBMSError",
+
+    # Examples: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/3170/
+    /Unable to find SOAP operation:/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/4713/
+    /Connection refused - connect\(2\) for "localhost" port 10001/ => "ECONNREFUSEDVBMSError",
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: this one on localhost port 10001 is an envoy error yes? I.e. the problem is our local forward proxy is down, not that VBMS is down.

File: app/exceptions/vbms_error.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/2557/
+    /Connection reset by peer - SSL_connect/ => "ECONNRESETVBMSError"
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: I'd consider this a transient error.

File: app/exceptions/vbms_error.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    /HTTPClient::ReceiveTimeoutError: exection expired/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/2322/
+    /HTTPClient::SendTimeoutError: exection expired/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/4708/
+    /HTTPClient::KeepAliveDisconnected:/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/2847/
+    /HTTPClient::ConnectTimeoutError: exection expired/ => "TransientVBMSError",
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: is that typo in the error message itself? or just here? should be *execution*

File: app/exceptions/vbms_error.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /Unable to find SOAP operation:/ => "TransientVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/4713/
+    /Connection refused - connect\(2\) for "localhost" port 10001/ => "ECONNREFUSEDVBMSError",
+
+    # Example: https://sentry.ds.va.gov/department-of-veterans-affairs/efolder/issues/2557/
+    /Connection reset by peer - SSL_connect/ => "ECONNRESETVBMSError"
+  }.freeze
+end
+# Many BGS calls fail in off-hours because BGS has maintenance time. These errors are classified
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: comment refers to BGS

File: app/exceptions/dependency_error.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+# Parent wrapper class for dependencies errors
+class DependencyError < StandardError
```
Comment (pkarman) [inline app/exceptions/dependency_error.rb]: I really like this refactor! we should try the same approach in Caseflow when we find time.

File: app/exceptions/vbms_error.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+# as transient errors and we ignore them in our reporting tools.
+
+class TransientVBMSError < VBMSError; end
+class ECONNREFUSEDVBMSError < VBMSError; end
```
Comment (pkarman) [inline app/exceptions/vbms_error.rb]: a nit about naming: I find it easier to scan names in Slack and Sentry when they have a common prefix and a more unique namespace at the end. So e.g. `VBMSError::Transient` or `VBMSError::ConnectionRefused`

File: config/environments/development.rb [hunk 10]
```diff
@@ -60,7 +60,7 @@
   ENV["METRICS_PASSWORD"] = "caseflow"
 
   config.sqs_create_queues = true
-  config.sqs_endpoint = 'http://localhost:4576'
+  config.sqs_endpoint = 'http://localhost:14576'
```
Comment (pkarman) [inline config/environments/development.rb]: are we running a sqs server locally?

Comment (nemodjuric) [reply config/environments/development.rb]: Yes, similarly to what we do in Caseflow as part of the localstack docker image. Also, in order to run tests locally we have to change settings in database.yml and secrets.yml currently, but it passes on Travis.

File: app/exceptions/vbms_error.rb (hunk 3 above)
Comment (nemodjuric) [reply app/exceptions/vbms_error.rb]: I fixed typos and took your naming suggestion.

File: config/environments/development.rb (hunk 10 above)
Comment (pkarman) [reply config/environments/development.rb]: I think if you create a `.env` file locally you can set the correct ports for your local env.

Alternately, we could make the defaults work locally and set the env vars for travis.

**Pull request merged**